edb/edgeql/parser/parser.py

```diff
@@ -36,12 +43,144 @@ def get_exception(self, native_err, context, token=None):
+                                continue
+
+                            elif isinstance(prevel, gr_commondl.OptExtending):
+                                # This is some SDL/DDL
+                                rule = 'definition'
+                            elif (
+                                isinstance(prevel, gr_exprs.Expr) or
+                                (
+                                    isinstance(prevel, tokens.T_COLON) and
+                                    isinstance(self.parser._stack[-3 - i][0],
+                                               gr_exprs.ShapePointer)
+                                )
+                            ):
+                                # This is some kind of shape.
+                                rule = 'shape'
+                            break
+                        elif isinstance(el, tokens.T_LPAREN):
+                            if need_match and isinstance(ltok,
+                                                         tokens.T_RPAREN):
+                                # This is matched, while we're looking
+                                # for unmatched parentheses.
+                                need_match = False
+                                continue
+                            # This could be an argument list or a tuple.
+                            elif isinstance(prevel, gr_exprs.NodeName):
+                                rule = 'list of arguments'
+                            else:
+                                rule = 'tuple'
+                            break
+                        elif isinstance(el, tokens.T_LBRACKET):
+                            if need_match and isinstance(ltok,
+                                                         tokens.T_RBRACKET):
+                                # This is matched, while we're looking
+                                # for unmatched brackets.
+                                need_match = False
+                                continue
+                            # This is either an array literal or
+                            # array index.
+                            elif isinstance(prevel, gr_exprs.Expr):
+                                rule = 'array slice'
+                            else:
+                                rule = 'array'
+                            break
+
+                if not token or token_kind == 'EOF':
                     msg = 'Unexpected end of line'
+                elif (
+                    rule == 'shape' and
+                    token_kind == 'IDENT' and
+                    isinstance(ltok, parsing.Nonterm)
+                ):
+                    # Make sure that the previous element in the stack
+                    # is some kind of Nonterminal, because if it's
+                    # not, this is probably not an issue of a missing
+                    # COMMA.
+                    msg = (f"It appears that a ',' is missing in {a(rule)} "
```

Comment:
We start error messages with a lowercase letter, because they are often prefixed with "Error: ".  Also, why does this have to be an error message and not a hint?